Add redundant type annotations lint

AlessioC31 opened the pull request:
Hello, I'm trying to add the `redundat_type_annotations` lint.

It's still WIP but I'd like to start gathering some feedbacks to be sure that I'm not doing things 100% wrong :)

Right now it still misses lints like:

- [x] `let foo: u32 = 5_u32`,
- [x] `let foo: String = STest2::func()`
- [x] `let foo: String = self.func()` (`MethodCall`)
- [x] refs
- [ ] Generics

I've some problems regarding the second example above, in the `init` part of the `Local` I have:

```rust
init: Some(
                Expr {
                    hir_id: HirId(DefId(0:24 ~ playground[e1bd]::main).58),
                    kind: Call(
                        Expr {
                            hir_id: HirId(DefId(0:24 ~ playground[e1bd]::main).59),
                            kind: Path(
                                TypeRelative(
                                    Ty {
                                        hir_id: HirId(DefId(0:24 ~ playground[e1bd]::main).61),
                                        kind: Path(
                                            Resolved(
                                                None,
                                                Path {
                                                    span: src/main.rs:77:21: 77:27 (#0),
                                                    res: Def(
                                                        Struct,
                                                        DefId(0:17 ~ playground[e1bd]::STest2),
                                                    ),
                                                    segments: [
                                                        PathSegment {
                                                            ident: STest2#0,
                                                            hir_id: HirId(DefId(0:24 ~ playground[e1bd]::main).60),
                                                            res: Def(
                                                                Struct,
                                                                DefId(0:17 ~ playground[e1bd]::STest2),
                                                            ),
                                                            args: None,
                                                            infer_args: true,
                                                        },
                                                    ],
                                                },
                                            ),
                                        ),
                                        span: src/main.rs:77:21: 77:27 (#0),
                                    },
                                    PathSegment {
                                        ident: get_numb#0,
                                        hir_id: HirId(DefId(0:24 ~ playground[e1bd]::main).62),
                                        res: Err,
                                        args: None,
                                        infer_args: true,
                                    },
                                ),
                            ),
                            span: src/main.rs:77:21: 77:37 (#0),
                        },
                        [],
                    ),
                    span: src/main.rs:77:21: 77:39 (#0),
                },
            ),
```

And I'm not sure how to get the return type of the function `STest2::func()` since the resolved path `DefId` points to the struct itself and not the function. Do you have any idea on how I could get this information in this case?

Thanks!

changelog: changelog: [`redundant_type_annotations`]: New lint to warn on redundant type annotations

fixes #9155

xFrednet commented on tests/ui/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,67 @@
+    5
+}
+
+fn main() {
```
I would suggest a few more tests. It's fine if not all of them work right away, but I definitely want to prevent false positives (FPs)

```rs
#[derive(Debug, Default)]
struct Cake<T> {
    _data: T,
}

fn make_something<T: Default>() -> T {
    T::default()
}

fn make_cake<T: Default>() -> Cake<T> {
    Cake::<T>::default()
}

fn test_generics() {
    // The type annotation is needed to determine T
    let _c: Cake<i32> = make_something();

    // The type annotation is needed to determine the topic
    let _c: Cake<u8> = make_cake();
}

fn test_non_locals() {
    // This shouldn't lint
    fn _arg(x: u32) -> u32 {
        x
    }

    // This could lint, but probably shouldn't
    let _closure_arg = |x: u32| x;
}

fn test_complex_types() {
    // Shouldn't lint, since the literal will be i32 otherwise
    let _u8: u8 = 128;

    // Should lint
    let _tuple_i32: (i32, i32) = (12, 13);

    // Shouldn't lint, since the tuple will be i32 otherwise
    let _tuple_u32: (u32, u32) = (1, 2);

    // Should lint, since the type is determined by the init value
    let _tuple_u32: (u32, u32) = (3_u32, 4_u32);

    // Should lint
    let _array: [i32; 3] = [5, 6, 7];

    // Shouldn't lint
    let _array: [u32; 2] = [8, 9];
}
```

I also feel like it would be better, to split up the main method into functions, which specify if and why an example should be linted :)

xFrednet commented on tests/ui/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,67 @@
+
+    let a: u32 = u32::MAX;
+
+    let a: u32 = ATest2::get_num(); // this should lint but doesn't (DefId refers to ATest2 and not to the func)
```
You can take a look at [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def) with this one. Maybe search in Clippy for examples how that is used to resolve the called function :)

AlessioC31 replied:
Thanks, this was very useulf, I managed to solve this problem :D

xFrednet commented on clippy_lints/src/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,143 @@
+
+impl LateLintPass<'_> for RedundantTypeAnnotations {
+    fn check_local<'tcx>(&mut self, cx: &LateContext<'tcx>, local: &'tcx rustc_hir::Local<'_>) {
+        if_chain! {
```
Here we should also include a check if the code comes from a macro. I believe it should be:

```rs
local.span.from_expansion()
```

or something close to that :)

AlessioC31 replied:
Done, thanks

xFrednet reviewed (commented):
Welcome to Clippy. Thank you for your patience while waiting on a review!

I've left some review comments. Please don't be discouraged by them, I believe this lint can become quite complex depending on how detailed and how many edge cases we want to catch. Please reach out if you require any further assistance. :)

AlessioC31 commented on clippy_lints/src/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,143 @@
+    #[clippy::version = "1.70.0"]
+    pub REDUNDANT_TYPE_ANNOTATIONS,
+    pedantic,
+    "Warns about needless / redundant type annotations."
```
Thanks!

AlessioC31 commented on clippy_lints/src/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,143 @@
+            false
+        },
+        // let a: String = String::new();
+        hir::QPath::TypeRelative(func_hir_ty, _) => {
```
Uhm in which cases do you think it could create I problem? I added your tests but I don't see anything weird happening, what am I missing?

xFrednet replied:
`QPaths` are always a bit complicated. This comment was a question if you've considered the case and came to the conclusion that the second argument isn't needed, or if there was just no test case that needed that.

Leaving it blank should be fine, I can't think of a test example for this.

AlessioC31 commented:
> Welcome to Clippy. Thank you for your patience while waiting on a review!
> 
> I've left some review comments. Please don't be discouraged by them, I believe this lint can become quite complex depending on how detailed and how many edge cases we want to catch. Please reach out if you require any further assistance. :)

Thanks for the review!

xFrednet commented:
> Thanks for the review!

You're welcome. Are you still working on the implementation or is the branch ready for another review? :)

AlessioC31 commented:
> > Thanks for the review!
> 
> You're welcome. Are you still working on the implementation or is the branch ready for another review? :)

I'm still working on it. Right now I've problems with `Lit` type of expressions (like how to map a `Lit` to a `Ty`) :)
Do you have any suggestion on where I could have a look?

bors commented:
:umbrella: The latest upstream changes (presumably #10601) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

xFrednet commented:
> I'm still working on it. Right now I've problems with `Lit` type of expressions (like how to map a `Lit` to a `Ty`) :) Do you have any suggestion on where I could have a look?

I don't think there is an easy way to map it. You could request the type of the expression, but that will probably not help too much.

> Or when is a type annotation useful in case of a `Lit`?

A type annotation is useful or even required when the literal is unsuffixed. For example `8` can be several types, the compiler would then look at its usages to determine the type. If the user writes `let x: u8 = 8`, the type should be kept IMO. However, if the literal has a suffix `8_u8` then the type can be removed. You can take a look at [`LitKind::is_unsuffixed()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.LitKind.html#method.is_unsuffixed) for this one. For `&str` etc it should be fine to remove the type as the literal is clear.

xFrednet commented on tests/ui/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,151 @@
+    // The type annotation is needed to determine the topic
+    let _c: Cake<u8> = make_cake();
+
+    // This should lint (doesn't)
+    let _c: Cake<u8> = make_cake::<u8>();
+
+    // This should lint (doesn't)
+    let _c: u8 = make_something::<u8>();
```
These are false negatives, which are fine to have in my opinion. I'd suggest merging the lint as is, we can always improve it later. Though, the comment should make it a bit clearer then. Same for the cases below.

Maybe:

```suggestion
    // This could be lint, but currently isn't 
    let _c: Cake<u8> = make_cake::<u8>();

    // This could be lint, but currently isn't
    let _c: u8 = make_something::<u8>();
```

xFrednet commented on tests/ui/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,151 @@
+
+fn main() {}
+
+// TODO: test refs
```
Could you address this TODO :)

xFrednet commented on tests/ui/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,151 @@
+
+    let _return: u32 = new_pie.return_an_int();
+
+    let _return: String = String::from("test");
```
This currently doesn't trigger a lint, even if the function comment says it should. I'm not quite sure why, as `String::new()` etc are covered.

AlessioC31 replied:
Wow, I missed that, thanks!

I digged into it a bit, it turns out it is related to generics.
When I do `cx.tcx.type_of(func_defid)` of the function, I get `EarlyBinder(fn(T) -> Self {<Self as std::convert::From<T>>::from})`

If I then execute `subst_identity` or `skip_binder` on this, I get `fn(T) -> Self {<Self as std::convert::From<T>>::from}`

Then, when I do `func_ty.fn_sig(cx.tcx).output()` where `func_ty` is the type I wrote previously, I get `Binder(Self, [])`, then I interpret the return type as `Self` instead of `String`.

Do you know what I could use in this case? I tried having a look at other lints but I was not able to find anything, thanks!

AlessioC31 replied:
cc @xFrednet :D

xFrednet replied:
Sorry, I missed your question earlier :sweat_smile:. I'm not totally sure about this one, but you might want to take a look at:
* [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def) to resolve the method target
* Maybe [qpath_res](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.qpath_res) to resolve `Self`?

If both of these don't work, I'd probably ask on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) the others might have a good idea for this one :)

AlessioC31 replied:
> Sorry, I missed your question earlier 😅. I'm not totally sure about this one, but you might want to take a look at:
> 
> * [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def) to resolve the method target
> * Maybe [qpath_res](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.qpath_res) to resolve `Self`?
> 
> If both of these don't work, I'd probably ask on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) the others might have a good idea for this one :)

I've tried the two methods you suggested without any luck. I've asked on Zulip, let's see.

Thanks!

xFrednet reviewed (commented):
Generally speaking, I think this is a good start, which we could actually merge as a minimum viable product. There are still several examples that could be improved. However, most of them seem a bit complicated to fix and the question is how much benefit we'll get from them. It should be fine as is, since all ToDos are false negatives. 

What are your thoughts? Do you want to try perfecting it in this PR or get this merge ready and create follow-ups? (Assuming, you're still interested in the lint)

AlessioC31 commented:
> Generally speaking, I think this is a good start, which we could actually merge as a minimum viable product. There are still several examples that could be improved. However, most of them seem a bit complicated to fix and the question is how much benefit we'll get from them. It should be fine as is, since all ToDos are false negatives.
> 
> What are your thoughts? Do you want to try perfecting it in this PR or get this merge ready and create follow-ups? (Assuming, you're still interested in the lint)

Hello @xFrednet , thanks again.

I'm still interested and I plan to perfecting this PR (at least adding support for primitive types and the suggestions you mentioned in the previous reviews).
I had to pause working on this because I had a busy period, it should be better next week!

xFrednet commented:
Sounds good, that's no problem at all, take your time! :)

bors commented:
:umbrella: The latest upstream changes (presumably #10578) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #10716) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

AlessioC31 commented:
> Hey @AlessioC31, what is the current status of the PR? Are there still open TODOs, besides the rebase? :)

Hello @xFrednet , there's still the problem with the `String::from` (and other generics since it's the same I guess). I've asked on Zulip but I haven't received any answer, maybe I can wrap this PR, I'll add support for refs, rebase and then it's ready to being merged. It won't lint in case generics are used but maybe I can come back at this when I understand how to do it

xFrednet commented:
> Hello @xFrednet, I think this is ready for review.
> 
> We miss a couple of things which I think they could be added in the future, like:
> 
>     * Support for generics

As long as we don't have false positives and only false negatives, that's totally fine :). Could you add this limitation, as well as the complex types and path one into the lint documentation? I like to use a `### Limitations` section to list such things. After this PR, we should ideally also create issues for these :)

>     * Refs returned from anything else than a `MethodCall` (maybe it's because of my little experience with rust yet, but I think this shouldn't happen often, what do you think?)

We should make sure that normal function calls are also detected. For one, because they're used quite often and because it should be a simple addition. I can mark the required changes in the next review :)

>     * Complex types (tuples, arrays, etc...)

Complex types are sadly complex to lint. I've been contemplating, how to best handle lints for these, for a different project. I sadly haven't found a good option. This is to say, that it's fine if these are not caught, probably. I guess that Clippy has quite a few lints that can't handle these :)

>     * `Path` to something different than a primitive type. Like for now this lint supports cases like `let a: u32 = u32::MAX` but it doesn't support things like `let a: MyStruct = Somewhere::MyStaticInstance`. Do you think we need to add this?

I believe that should be fine. Such a pattern is rarely used in the projects I'm using. We can leave this as an improvement for the future :)

> Thanks!

You're very welcome, thank you for working on Clippy! I'll review this PR in the upcoming week :)

bors commented:
:umbrella: The latest upstream changes (presumably #10810) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

xFrednet commented on clippy_lints/src/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,201 @@
+    /// Code without type annotations is shorter and in most cases
+    /// more idiomatic and easier to modify.
+    ///
+    /// ### Example
```
It would be good to have a `### Limitations` section, as suggested in this comment:

https://github.com/rust-lang/rust-clippy/pull/10570#issuecomment-1556238326

AlessioC31 replied:
Done thanks!

AlessioC31 commented on clippy_lints/src/redundant_type_annotations.rs:
```diff
@@ -0,0 +1,201 @@
+        {
+            match &init.kind {
+                // When the initialization is a call to a function
+                hir::ExprKind::Call(init_call, _) => {
```
Yes, what I meant in the previous comment was that I don't think it will handle refs returned from function calls

xFrednet replied:
Ahh okay, that's totally fine :)

AlessioC31 commented:
I think I've addressed all your comments and rebased to `master`. Should I squash the commits?

Thanks!

xFrednet commented:
> I think I've addressed all your comments and rebased to `master`.

There still seems to be a conflict, could you resolve that one?

> Should I squash the commits?

It's not mandatory, but would be nice, 1 - 5 commits would be a good balance :)

I intend to review this over the weekend :)

AlessioC31 commented:
> > I think I've addressed all your comments and rebased to `master`.
> 
> There still seems to be a conflict, could you resolve that one?
> 
> > Should I squash the commits?
> 
> It's not mandatory, but would be nice, 1 - 5 commits would be a good balance :)
> 
> I intend to review this over the weekend :)

Ok I've rebased to master and reduced the amount of commits, thanks!

AlessioC31 commented:
Rebased again :D

bors commented:
:pushpin: Commit a9b468f2b51a251d0d718059299d0ccd0e561389 has been approved by `xFrednet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=xFrednet a9b468f2b51a251d0d718059299d0ccd0e561389 -->
<!-- homu: {"type":"Approved","sha":"a9b468f2b51a251d0d718059299d0ccd0e561389","approver":"xFrednet","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/5193389549/jobs/9363893454), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/5193389545/jobs/9363915887), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/5193389548/jobs/9364147340)
Approved by: xFrednet
Pushing cc8ead2cce5a99ac8d4e963570c7eb6ed18f66b7 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"xFrednet","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/5193389548/jobs/9364147340","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/5193389549/jobs/9363893454","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/5193389545/jobs/9363915887"},"merge_sha":"cc8ead2cce5a99ac8d4e963570c7eb6ed18f66b7"} -->
